// cmd-ts/dist/esm/multioption.js
import { findOption } from './newparser/findOption';
import * as Result from './Result';
import chalk from 'chalk';
/**
 * Like `option`, but can accept multiple options, and expects a decoder from a list of strings.
 * An error will highlight all option occurences.
 */
export function multioption(config) {
    return {
        helpTopics() {
            var _a, _b, _c, _d;
            const displayName = (_a = config.type.displayName) !== null && _a !== void 0 ? _a : 'value';
            let usage = `--${config.long} <${displayName}>`;
            if (config.short) {
                usage += `, -${config.short}=<${displayName}>`;
            }
            const defaults = [];
            const defaultValueFn = (_b = config.defaultValue) !== null && _b !== void 0 ? _b : config.type.defaultValue;
            if (defaultValueFn) {
                try {
                    const defaultValue = defaultValueFn();
                    if ((_c = config.defaultValueIsSerializable) !== null && _c !== void 0 ? _c : config.type.defaultValueIsSerializable) {
                        defaults.push('default: ' + chalk.italic(defaultValue));
                    }
                    else {
                        defaults.push('[...optional]');
                    }
                }
                catch (e) { }
            }
            return [
                {
                    category: 'options',
                    usage,
                    defaults,
                    description: (_d = config.description) !== null && _d !== void 0 ? _d : 'self explanatory',
                },
            ];
        },
        register(opts) {
            opts.forceOptionLongNames.add(config.long);
            if (config.short) {
                opts.forceOptionShortNames.add(config.short);
            }
        },
        async parse({ nodes, visitedNodes, }) {
            var _a, _b;
            const options = findOption(nodes, {
                longNames: [config.long],
                shortNames: config.short ? [config.short] : [],
            }).filter((x) => !visitedNodes.has(x));
            const defaultValueFn = (_a = config.defaultValue) !== null && _a !== void 0 ? _a : config.type.defaultValue;
            if (options.length === 0 && typeof defaultValueFn === 'function') {
                try {
                    return Result.ok(defaultValueFn());
                }
                catch (e) {
                    const message = `Failed to resolve default value for '--${config.long}': ${e.message}`;
                    return Result.err({
                        errors: [
                            {
                                nodes: [],
                                message,
                            },
                        ],
                    });
                }
            }
            for (const option of options) {
                visitedNodes.add(option);
            }
            const optionValues = [];
            const errors = [];
            const flagNodes = [];
            for (const option of options) {
                const providedValue = (_b = option.value) === null || _b === void 0 ? void 0 : _b.node.raw;
                if (providedValue === undefined) {
                    flagNodes.push(option);
                    continue;
                }
                optionValues.push(providedValue);
            }
            if (flagNodes.length > 0) {
                errors.push({
                    nodes: flagNodes,
                    message: `Expected to get a value, found a flag`,
                });
            }
            if (errors.length > 0) {
                return Result.err({ errors });
            }
            const multiDecoded = await Result.safeAsync(config.type.from(optionValues));
            if (Result.isErr(multiDecoded)) {
                return Result.err({
                    errors: [{ nodes: options, message: multiDecoded.error.message }],
                });
            }
            return multiDecoded;
        },
    };
}
